objects/src/block/mod.rs

```diff
@@ -146,26 +152,41 @@ impl Block {
     // --------------------------------------------------------------------------------------------
 
     fn validate(&self) -> Result<(), BlockError> {
+        let account_count = self.updated_accounts.len();
+        if account_count > MAX_ACCOUNTS_PER_BLOCK {
+            return Err(BlockError::TooManyAccountsUpdated(account_count));
+        }
+
         let batch_count = self.output_note_batches.len();
         if batch_count > MAX_BATCHES_PER_BLOCK {
             return Err(BlockError::TooManyTransactionBatches(batch_count));
         }
 
-        for batch in self.output_note_batches.iter() {
-            if batch.len() > MAX_NOTES_PER_BATCH {
-                return Err(BlockError::TooManyNotesInBatch(batch.len()));
-            }
+        // We can't check input notes here because they're not stored in the block,
+        // so we check that nullifier count is not bigger than maximum input notes per block.
+        let nullifier_count = self.nullifiers.len();
+        if nullifier_count > MAX_INPUT_NOTES_PER_BLOCK {
+            return Err(BlockError::TooManyNullifiersInBlock(nullifier_count));
         }
 
-        let mut notes = BTreeSet::new();
+        let mut output_notes = BTreeSet::new();
+        let mut output_note_count = 0;
         for batch in self.output_note_batches.iter() {
+            if batch.len() > MAX_OUTPUT_NOTES_PER_BATCH {
+                return Err(BlockError::TooManyNotesInBatch(batch.len()));
+            }
+            output_note_count += batch.len();
             for note in batch.iter() {
-                if !notes.insert(note.id()) {
+                if !output_notes.insert(note.id()) {
                     return Err(BlockError::DuplicateNoteFound(note.id()));
                 }
             }
         }
 
+        if output_note_count > MAX_OUTPUT_NOTES_PER_BLOCK {
+            return Err(BlockError::TooManyNotesInBlock(output_note_count));
+        }
```

Comment:
I think this is redundant. We check the right number of batches and the right number of output notes par batch above - so, this should always pass, no?

Reply:
It's true for now just because we calculate `MAX_OUTPUT_NOTES_PER_BLOCK` as `MAX_OUTPUT_NOTES_PER_BATCH * MAX_BATCHES_PER_BLOCK` for simplicity. But it is maximum theoretically possible value, so in the future we might want to reduce this number.